File: platforms/android/java/app/msime/client/candidate/OnlineCandidatePolicy.java
package app.msime.client;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/**
 * Bounds and identifies asynchronous cloud and AI results before they return to Engine.
 *
 * <p>Both providers answer a query Engine has usually moved past by the time the network replies.
 * The signature is what stops a second request for a state already asked about, and the byte and
 * shape limits are what stop a service reply from reaching Engine as anything other than a short
 * list of plain candidate strings.
 *
 * <p>The rules live here without any JSON or Android dependency so the host test suite can run
 * them; reading the service envelopes stays in the input service, where JSON already lives.
 */
public final class OnlineCandidatePolicy {
    /** How long a composition has to hold still before either provider is asked. */
    public static final long QUIET_INTERVAL_MILLIS = 350;
    public static final int MAX_CLOUD_RESPONSE_BYTES = 256 * 1024;
    public static final int MAX_AI_RESPONSE_BYTES = 1024 * 1024;
    public static final int MAX_AI_CONTENT_BYTES = 64 * 1024;
    private static final int MAX_CANDIDATE_BYTES = 4096;
    private static final int MAX_CANDIDATE_LIMIT = 10;

    private OnlineCandidatePolicy() {}

    /**
     * Identity of one online request.
     *
     * <p>{@code assistant} is the AI configuration document when the assistant is enabled and
     * empty otherwise: changing model or prompt has to ask again for the same composition, while a
     * query that only advanced its generation without changing the cache key must not, and a
     * disabled assistant's settings are not going to be consulted either way.
     */
    public static String signature(long sessionId, String cacheKey, String identity,
            boolean cloudCandidates, String assistant) {
        return "session=" + sessionId + "|cache=" + field(cacheKey)
            + "|identity=" + field(identity) + "|cloud=" + cloudCandidates
            + "|assistant=" + field(assistant);
    }

    /** Whether the cloud provider should be asked for this query. */
    public static boolean requestsCloud(boolean cloudCandidates, boolean cloudEligible) {
        return cloudCandidates && cloudEligible;
    }

    /** Whether the AI provider should be asked for this query. */
    public static boolean requestsAi(boolean aiEligible, boolean assistantEnabled) {
        return aiEligible && assistantEnabled;
    }

    /** The configured candidate limit, or zero when it is outside what the shared host accepts. */
    public static int aiCandidateLimit(int limit) {
        return limit >= 1 && limit <= MAX_CANDIDATE_LIMIT ? limit : 0;
    }

    /** Whether a cloud body is small enough to hand to the shared parser. */
    public static boolean acceptsCloudBody(String body) {
        return body != null && !body.isEmpty() && utf8Length(body) <= MAX_CLOUD_RESPONSE_BYTES;
    }

    /** Whether an AI reply is small enough to parse. */
    public static boolean acceptsAiBody(String body) {
        return body != null && !body.isEmpty() && utf8Length(body) <= MAX_AI_RESPONSE_BYTES;
    }

    /** Whether the JSON-mode content inside an AI reply is small enough to parse. */
    public static boolean acceptsAiContent(String content) {
        return content != null && !content.isEmpty() && utf8Length(content) <= MAX_AI_CONTENT_BYTES;
    }

    /**
     * The candidates a reply may contribute: provider order, no duplicates, capped at the limit.
     *
     * <p>Blank, control-bearing and over-long entries are skipped rather than failing the batch,
     * matching the shared parser: one unusable entry does not discard the usable ones beside it.
     */
    public static List<String> aiCandidates(List<String> texts, int limit) {
        List<String> result = new ArrayList<>();
        if (texts == null || aiCandidateLimit(limit) == 0) return result;
        for (String text : texts) {
            if (result.size() == limit) break;
            if (text == null || text.trim().isEmpty() || utf8Length(text) > MAX_CANDIDATE_BYTES
                    || hasControl(text) || result.contains(text)) {
                continue;
            }
            result.add(text);
        }
        return result;
    }

    private static String text(String value) { return value == null ? "" : value; }

    private static String field(String value) {
        value = text(value);
        return value.length() + ":" + value;
    }

    private static boolean hasControl(String value) {
        for (int index = 0; index < value.length();) {
            int codePoint = value.codePointAt(index);
            if (codePoint <= 0x1f || (codePoint >= 0x7f && codePoint <= 0x9f)) return true;
            index += Character.charCount(codePoint);
        }
        return false;
    }

    private static int utf8Length(String value) {
        return value.getBytes(StandardCharsets.UTF_8).length;
    }
}

File: platforms/harmony/entry/src/main/ets/keyboard/candidate/OnlineCandidatePolicy.ts
import { utf8Length } from '../Utf8';

export interface OnlineAssistantConfig {
  enabled: boolean;
  model: string;
  candidate_limit: number;
}

export interface OnlineQuery {
  generation: number;
  session_id: number;
  identity: string;
  cache_key: string;
  cloud_eligible: boolean;
  ai_eligible: boolean;
  cloud_candidates: boolean;
  ai_assistant?: OnlineAssistantConfig | null;
}

interface AiMessage {
  content?: string;
}

interface AiChoice {
  message?: AiMessage;
}

interface AiEnvelope {
  choices?: AiChoice[];
  error?: Object | null;
}

interface AiCandidate {
  text?: string;
}

interface AiContent {
  candidates?: AiCandidate[];
}

/** Bounds and identifies asynchronous cloud/AI results before they return to Engine. */
export class OnlineCandidatePolicy {
  static readonly QUIET_INTERVAL_MS: number = 350;
  static readonly MAX_CLOUD_RESPONSE_BYTES: number = 256 * 1024;
  static readonly MAX_AI_RESPONSE_BYTES: number = 1024 * 1024;

  /** Whether a cloud reply is bounded before it is handed to the native parser. */
  static acceptsCloudBody(body: string | null | undefined): boolean {
    return body !== null && body !== undefined && body.length > 0
      && utf8Length(body) <= OnlineCandidatePolicy.MAX_CLOUD_RESPONSE_BYTES;
  }

  static signature(query: OnlineQuery): string {
    const assistant: OnlineAssistantConfig | null | undefined = query.ai_assistant;
    return `${query.session_id}:${query.cache_key}:${query.identity}:`
      + `${query.cloud_candidates}:${assistant?.enabled === true ? JSON.stringify(assistant) : ''}`;
  }

  static aiCandidates(body: string, limit: number): string[] | null {
    if (utf8Length(body) > OnlineCandidatePolicy.MAX_AI_RESPONSE_BYTES
      || !Number.isInteger(limit) || limit < 1 || limit > 10) {
      return null;
    }
    try {
      const envelope: AiEnvelope = JSON.parse(body) as AiEnvelope;
      if (envelope.error !== undefined && envelope.error !== null
        || envelope.choices === undefined || envelope.choices.length === 0) {
        return null;
      }
      const content: string | undefined = envelope.choices[0].message?.content;
      if (content === undefined || utf8Length(content) > 64 * 1024) return null;
      const document: AiContent = JSON.parse(content) as AiContent;
      if (document.candidates === undefined || !Array.isArray(document.candidates)) return null;
      const result: string[] = [];
      for (const candidate of document.candidates) {
        const text: string | undefined = candidate.text;
        if (text === undefined || text.trim().length === 0 || utf8Length(text) > 4096
          || OnlineCandidatePolicy.hasControl(text)) {
          continue;
        }
        if (!result.includes(text)) result.push(text);
        if (result.length === limit) break;
      }
      return result;
    } catch {
      return null;
    }
  }

  private static hasControl(value: string): boolean {
    for (const character of value) {
      const code: number = character.codePointAt(0) ?? 0;
      if (code <= 0x1f || (code >= 0x7f && code <= 0x9f)) return true;
    }
    return false;
  }
}
